### PR TITLE
Initial Royal Jelly Port Upstream

### PR DESCRIPTION
**Pull request by The-Aerec**
**_I affirm:_**
- [x] I understand that if I do not agree to the following points by completing the checkboxes my PR will be ignored.
- [x] I understand I should leave resolving conversations to the LandSandBoat team so that reviewers won't miss what was said.
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/LandSandBoat/server/blob/base/CODE_OF_CONDUCT.md).
- [x] I have _**tested my code and the things my code has changed**_ since the last commit in the PR and will test after any later commits.

## What does this pull request do?

Pushes missing Lua Files for Royal Jelly BCNM Upstream from ASB for comments

## Steps to test these changes

1. !additem 1131 (Star Orb)
2. !zone Waughroon Shrine
3. Enter BCNM 
4. Test mechanics for princess jellys slowly moving into room
5. Test mechanics for Jellys combining if not killed before they meet
6. Test win if all princess jelly killed
7. Test win if queen jelly is killed

### CODE REVIEW

File: scripts/globals/bcnm.lua [hunk 1]
```diff
@@ -275,7 +275,7 @@ local battlefields =
     --  { 10,   74, 1551 },   -- Crustacean Conundrum (BS20) -- TODO: You can only do 0-2 damage no matter what your attack is
         { 11,   75, 1552 },   -- Grove Guardians (BS30)
     --  { 12,   76, 1553 },   -- The Hills are Alive (KS99) -- TODO: Tartaruga Gigante is not coded
-    --  { 13,   77, 1131 },   -- Royal Jelly (BS40) -- TODO: all combat mechanics, loot
+        { 13,   77, 1131 },   -- Royal Jelly (BS40) -- TODO: all combat mechanics, loot
```
Comment (kaincenteno) [inline scripts/globals/bcnm.lua]: looks like you added the combat mechanic and loot, maybe that TODO needs to be removed?

also awesome  for contributing with this BCNM 😄  its one of my favourites 😄

Comment (zach2good) [inline scripts/globals/bcnm.lua]: TODO can go now

File: scripts/zones/Waughroon_Shrine/IDs.lua [hunk 2]
```diff
@@ -84,6 +84,13 @@ zones[xi.zone.WAUGHROON_SHRINE] =
             17367285,
         },
     },
+
+    royalJellyQueens =
```
Comment (zach2good) [inline scripts/zones/Waughroon_Shrine/IDs.lua]: These can be looked up at runtime with zone:queryEntitiesByName or GetFirstID taking advantage of the fact that they're all offset from eachother by 10. Or drop in GetTableOfIDs for a one-line replacement

File: scripts/zones/Waughroon_Shrine/bcnms/royal_jelly.lua [hunk 3]
```diff
@@ -28,10 +30,10 @@ battlefieldObject.onBattlefieldLeave = function(player, battlefield, leavecode)
     end
 end
 
-battlefieldObject.onEventUpdate = function(player, csid, option, npc)
+battlefieldObject.onEventUpdate = function(player, csid, option)
 end
 
-battlefieldObject.onEventFinish = function(player, csid, option, npc)
+battlefieldObject.onEventFinish = function(player, csid, option)
```
Comment (zach2good) [inline scripts/zones/Waughroon_Shrine/bcnms/royal_jelly.lua]: Can you add back in the removed npc arg

File: scripts/zones/Waughroon_Shrine/mobs/Princess_Jelly.lua [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+    mob:setMod(xi.mod.REGEN, 3)
+    mob:setLocalVar('mobElement', math.random(1, 8))
+    mob:addMod(mevaList[mob:getLocalVar('mobElement')][1], -250)
+    mob:addMod(mevaList[mob:getLocalVar('mobElement')][2], 1000)
+end
+
+local function getDistanceFromCenter(bfNum, mob)
+    local pos = mob:getPos()
+
+    local difX =  pos.x - centers[bfNum][1]
```
Comment (zach2good) [inline scripts/zones/Waughroon_Shrine/mobs/Princess_Jelly.lua]: Slight alignment tweak here

File: scripts/zones/Waughroon_Shrine/npcs/Armoury_Crate.lua [hunk 5]
```diff
@@ -404,6 +404,65 @@ local loot =
         },
     },
 
+    -- BCNM Royal Jelly
+    [77] =
+    {
+        {
+            { itemid = 637, droprate = 1000 }, -- Slime oil
```
Comment (zach2good) [inline scripts/zones/Waughroon_Shrine/npcs/Armoury_Crate.lua]: Can you add all these items to items.lua

**Review by zach2good — COMMENTED** [review]
Looking pretty close 👍

**Comment by The-Aerec** [comment]
Comments addressed

File: scripts/globals/bcnm.lua (hunk 1 above)
Comment (The-Aerec) [reply scripts/globals/bcnm.lua]: done

Comment (The-Aerec) [reply scripts/globals/bcnm.lua]: done

File: scripts/zones/Waughroon_Shrine/IDs.lua (hunk 2 above)
Comment (The-Aerec) [reply scripts/zones/Waughroon_Shrine/IDs.lua]: done

File: scripts/zones/Waughroon_Shrine/bcnms/royal_jelly.lua (hunk 3 above)
Comment (The-Aerec) [reply scripts/zones/Waughroon_Shrine/bcnms/royal_jelly.lua]: done

File: scripts/zones/Waughroon_Shrine/mobs/Princess_Jelly.lua (hunk 4 above)
Comment (The-Aerec) [reply scripts/zones/Waughroon_Shrine/mobs/Princess_Jelly.lua]: done

File: scripts/zones/Waughroon_Shrine/npcs/Armoury_Crate.lua (hunk 5 above)
Comment (The-Aerec) [reply scripts/zones/Waughroon_Shrine/npcs/Armoury_Crate.lua]: done

**Review by zach2good — APPROVED** [review]
LGTM 🚀 

(I've never played this on retail, so I don't know if it plays correctly. I assume since this has been running downstream and is quite popular that its vaguely tuned and correct and isn't a loot pinata)

**Comment by zach2good** [comment]
TODO: once someone else signs this off, I'll be the one to merge it because I want to fiddle with the PR title and commit message on the way in, etc.

**Pull request merged**